mcs/class/Mono.Posix/Mono.Unix/UnixMarshal.cs

```diff
@@ -325,27 +325,34 @@ public static IntPtr StringToHeap (string s, int index, int count, Encoding enco
 			if (encoding == null)
 				throw new ArgumentNullException ("encoding");
 
-			int min_byte_count = encoding.GetMaxByteCount(1);
-			char[] copy = s.ToCharArray (index, count);
-			byte[] marshal = new byte [encoding.GetByteCount (copy) + min_byte_count];
+			if (index < 0 || count < 0)
+				throw new ArgumentOutOfRangeException ((index < 0 ? "index" : "count"),
+					 "Non - negative number required.");
 
-			int bytes_copied = encoding.GetBytes (copy, 0, copy.Length, marshal, 0);
+			if (s.Length - index < count)
+				throw new ArgumentOutOfRangeException ("s", "Index and count must refer to a location within the string.");
 
-			if (bytes_copied != (marshal.Length-min_byte_count))
-				throw new NotSupportedException ("encoding.GetBytes() doesn't equal encoding.GetByteCount()!");
+			int null_terminator_count = encoding.GetMaxByteCount(1);
+			int length_without_null = encoding.GetByteCount(s);
+			int marshalLength = length_without_null + null_terminator_count;
```

Comment:
For better sanity, this should be:

```csharp
int marshalLength = checked (length_without_null + null_terminator_count);
```

We want to avoid integer overflow; that could be bad.